src/txn/txn_ckpt.c

```diff
@@ -631,29 +642,23 @@ __txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
 	tracking = true;
 
 	/*
-	 * Get a list of handles we want to flush; for named checkpoints this
-	 * may pull closed objects into the session cache.
-	 *
 	 * We want to skip checkpointing clean handles whenever possible.  That
 	 * is, when the checkpoint is not named or forced.  However, we need to
 	 * take care about ordering with respect to the checkpoint transaction.
 	 *
-	 * If we skip clean handles before starting the transaction, the
+	 * We can't skip clean handles before starting the transaction or the
 	 * checkpoint can miss updates in trees that become dirty as the
 	 * checkpoint is starting.  If we wait until the transaction has
 	 * started before locking a handle, there could be a metadata-changing
 	 * operation in between (e.g., salvage) that will cause a write
 	 * conflict when the checkpoint goes to write the metadata.
 	 *
-	 * First, gather all handles, then start the checkpoint transaction,
-	 * then release any clean handles.
+	 * Hold the schema lock while starting the transaction and gathering
+	 * handles so the set we get is complete and correct.
 	 */
-	WT_ASSERT(session, session->ckpt_handle_next == 0);
-	WT_WITH_SCHEMA_LOCK(session,
-	    WT_WITH_TABLE_READ_LOCK(session,
-		ret = __checkpoint_apply_all(
-		    session, cfg, __wt_checkpoint_get_handles, NULL)));
-	WT_ERR(ret);
+	__wt_spin_lock(session, &conn->schema_lock);
+	F_SET(session, WT_SESSION_LOCKED_SCHEMA);
+	locked_schema = true;
```

Comment:
If you end up holding the schema lock across this block of code, I'd prefer if it was pulled into a separate function so we can use a ``WT_WITH_LOCK`` macro. FWIW I think is reasonable to hold the schema lock here since it blocks other schema level operations (create, drop, rename) from updating the metadata while checkpoint is deciding which handles should be included in the checkpoint. It may not be strictly necessary, but there are enough different co-ordination mechanisms in play here, that holding the schema lock makes me feel more comfortable.